packages/abi-typegen/src/templates/dts.ts

```diff
@@ -0,0 +1,167 @@
+  }
+
+  return [
+    `export type ${structName}Input${typeAnnotations} = ${inputValues};`,
+    `export type ${structName}Output${typeAnnotations} = ${outputValues};`,
+  ].join('\n');
+}
+
+export function renderDtsEnumExportsTemplate(params: {
+  structName: string;
+  inputValues: string;
+  outputValues: string;
+}) {
+  const { structName } = params;
+
+  const inputValues = `Enum<{ ${params.inputValues} }>`;
+  let outputValues = `Enum<{ ${params.outputValues} }>`;
+
+  if (inputValues === outputValues) {
+    outputValues = `${structName}Input`;
+  }
+
+  return [
+    `export type ${structName}Input = ${inputValues};`,
+    `export type ${structName}Output = ${outputValues};`,
+  ].join('\n');
+}
+
+export function renderDtsCommonImportTemplate(params: { commonTypesInUse: string[] }) {
+  const { commonTypesInUse } = params;
+  return `import type { ${commonTypesInUse.join(', ')} } from "./common";`;
+}
+
+/*
+  Render method
+*/
+export function renderDtsTemplate(params: { abi: Abi }) {
+  const { name, types, functions, commonTypesInUse, fuelsVersion } = params.abi;
+
+  /*
+    First we format all attributes
+  */
+  const fnsTypedefs = functions.map((f) => f.getDeclaration());
+
+  const fnsFragments = functions.map((f) => `${f.attributes.name}: FunctionFragment;`);
+
+  const encoders = functions.map((f) =>
+    renderDtsEncoderTemplate({
+      functionName: f.attributes.name,
+      input: f.attributes.inputs,
+    })
+  );
+
+  const decoders = functions.map((f) =>
+    renderDtsDecoderTemplate({
+      functionName: f.attributes.name,
+    })
+  );
+
+  const structs = types
+    .filter((t) => t.name === 'struct')
+    .map((t) => {
+      const st = t as StructType; // only structs here
+      const structName = st.getStructName();
+      const inputValues = st.getStructContents({ types, target: TargetEnum.INPUT });
+      const outputValues = st.getStructContents({ types, target: TargetEnum.OUTPUT });
+      const typeAnnotations = st.getStructDeclaration({ types });
+      return renderDtsStructExportsTemplate({
+        structName,
+        typeAnnotations,
+        inputValues,
+        outputValues,
+      });
+    });
+
+  const enums = types
+    .filter((t) => t.name === 'enum')
+    .map((t) => {
+      const et = t as EnumType; // only enums here
+      const structName = et.getStructName();
+      const inputValues = et.getStructContents({ types, target: TargetEnum.INPUT });
+      const outputValues = et.getStructContents({ types, target: TargetEnum.OUTPUT });
+      return renderDtsEnumExportsTemplate({
+        structName,
+        inputValues,
+        outputValues,
+      });
+    });
+
+  // Handles custom common types
+  let commonImports = renderDtsCommonImportTemplate({ commonTypesInUse });
+
+  commonImports = commonTypesInUse.length ? `\n${commonImports}\n` : '';
+
+  // glue everything
+  const output = `/* Autogenerated file. Do not edit manually. */
```

Comment:
We could replace these templates by placing them on files and using `handlebars` or another template processor to replace the dynamic parts.